### ISSUE TITLE
Jaeger UI is slowly losing operations that are still available via the API

### ISSUE THREAD

**Issue by Azim-Palmer** [jaegertracing/jaeger#2420]
**Describe the bug**
I have Jaeger hooked up to Elasticsearch and have it running for over a month. Over time my operations count via the jaeger ui went up to 52, however after about a week I noticed it dropped down to 32 and a day later has risen to 46.

**To Reproduce**
1. Have Jaeger hooked up to Elasticsearch for over a month
2. Create a number of different spans over the course of that month
3. Wait several days (sorry I can't be more specific)
4. Observe the # of operations has dropped

**Expected behavior**
Every historical operation that has spans logged against it should be available via the UI

**Version (please complete the following information):**
 - OS: Linux
 - Jaeger version: 1.17
 - Deployment: Docker

**What troubleshooting steps did you try?**
Checked the Elasticsearch index and all the historical operations are there (and there's 100 of them)
Confirmed that the operations **aren't** present on the response of via /api/services/SERVICENAME/operations

I've confirmed that I **can** still get the spans via the api via /api/traces?end=1598868060000000&limit=200&lookback=custom&maxDuration&minDuration&operation=WebRequest%3A%2Fquotes%2F19&service=SERVICENAME&start=1588305660000000

**Additional context**
Just in case it's important, I'm using a const sampler at the moment so it's logging everything !

**Comment by jpkrohling** [comment]
>  Confirmed that the operations aren't present on the response of via /api/services/SERVICENAME/operations

This sounds like a problem in the ES backend, not UI.

**Comment by Azim-Palmer** [comment]
@jpkrohling How would I go about confirming that? 

I've "Checked the Elasticsearch index and all the historical operations are there (and there's 100 of them)" - to elaborate I created an index patten on "jaeger-service-*" and it's returning everything that I'm expecting to see.

I don't believe I've changed anything significant on the Elastic side, the instance is literally just for Jaeger.

**Comment by jpkrohling** [comment]
Sorry, I meant the ES-specific part of the Jaeger backend. Likely around here:

https://github.com/jaegertracing/jaeger/blob/66d0d20b98126c59ce687f4d6ccfe7c28e0fc4f1/plugin/storage/es/spanstore/reader.go#L247-L270

**Comment by goku321** [comment]
Want to take this one up.

**Comment by jpkrohling** [comment]
It's yours. Let us know if you need help.

**Comment by goku321** [comment]
Why are we subtracting 24 here?
https://github.com/jaegertracing/jaeger/blob/66d0d20b98126c59ce687f4d6ccfe7c28e0fc4f1/plugin/storage/es/spanstore/reader.go#L179

**Comment by yurishkuro** [comment]
I assume because indices are expected to be daily

**Comment by anshul35** [comment]
Can I work on this one ?

**Comment by goku321** [comment]
Hi @anshul35, feel free to work on this. I couldn't make much progress

**Comment by stillme1** [comment]
Hi I am absolutely beginner in open source, can someone please help me out here.

@jpkrohling  I went through some code, it seems like the issue is that we are just using the start time as currentTime - SpanReader.maxSpanAge which is why the API would be giving partial result, while the entire result is available in db.

https://github.com/jaegertracing/jaeger/blob/main/plugin/storage/es/spanstore/reader.go#L302

**Comment by lenikadali** [comment]
Wanted to work on this issue and wanted to update [this comment](https://github.com/jaegertracing/jaeger/issues/2420#issuecomment-681732834) since the relevant code has been moved.

The relevant lines are [297 to 326](https://github.com/jaegertracing/jaeger/blob/main/internal/storage/v1/elasticsearch/spanstore/reader.go#L297-L326) I think.

I am curious to know if the issue is related to this [`TODO`](https://github.com/jaegertracing/jaeger/blob/main/internal/storage/v1/elasticsearch/spanstore/reader.go#L317-L325). If not, given there is a test file [here](https://github.com/jaegertracing/jaeger/blob/main/internal/storage/v1/elasticsearch/spanstore/reader_test.go), how would we check that all the API operations are provided in the ES backend and will therefore show up on the UI.

**Comment by yurishkuro** [comment]
@lenikadali i don't think that specific code is at fault. Jaeger materializes services/operations into a separate index, which likely has independent code path for reading compared to reading spans, and it's possible there is a divergence there. The behavior described by the user sounds like jaeger is reading operations from a shorter time window than spans.

**Comment by madugushivakumar** [comment]
Thanks for the clarification @yurishkuro.

This PR addresses that divergence by extending the time window used when
querying the service/operation indices in `GetOperations`, so operations
are no longer constrained by the span lookback window.

Happy to adjust the approach or add tests if needed.
